jena-arq/src/main/java/org/apache/jena/riot/RDFParserBuilder.java

```diff
@@ -612,10 +612,17 @@ public RDFParser build() {
         if ( errorHandler$ == null )
             errorHandler$ = ErrorHandlerFactory.getDefaultErrorHandler();
 
-        if ( path != null && baseUri == null )
-            baseUri = IRILib.filenameToIRI(path.toString());
-        if ( path == null && baseUri == null && uri != null )
-            baseUri = uri;
+        if ( path != null ) {
+            if ( baseUri == null )
+                baseUri = IRILib.filenameToIRI(path.toString());
+        } else {
+            if ( baseUri == null && uri != null ) {
+                if ( resolver != null )
+                    baseUri = resolver.resolveToString(uri);
+                else
+                    baseUri = IRIResolver.resolveString(uri);
+            }
+        }
```

Comment:
Or maybe

```java
if ( baseUri == null ) {
    if ( path != null ) {
        baseUri = IRILib.filenameToIRI(path.toString());
    } else if ( uri != null ) {
        baseUri = resolver != null ? resolver.resolveToString(uri) : IRIResolver.resolveString(uri);
    }
}
```

I think it should eval the same way, but moving the common condition `baseUri == null` out?

Reply:
Yes, much better!